jax_cosmo/probes.py

```diff
@@ -18,21 +19,56 @@
 def weak_lensing_kernel(cosmo, pzs, z, ell):
     """
     Returns a weak lensing kernel
+
+    Note: this function handles differently nzs that correspond to extended redshift
+    distribution, and delta functions.
     """
     z = np.atleast_1d(z)
     zmax = max([pz.zmax for pz in pzs])
     # Retrieve comoving distance corresponding to z
     chi = bkgrd.radial_comoving_distance(cosmo, z2a(z))
 
-    @vmap
-    def integrand(z_prime):
-        chi_prime = bkgrd.radial_comoving_distance(cosmo, z2a(z_prime))
-        # Stack the dndz of all redshift bins
-        dndz = np.stack([pz(z_prime) for pz in pzs], axis=0)
-        return dndz * np.clip(chi_prime - chi, 0) / np.clip(chi_prime, 1.0)
+    # Extract the indices of pzs that can be treated as extended distributions,
```

Comment:
@dlanzieri I updated a little bit the logic for how to compute the kernel in the 2 different cases. Your implementation worked :-) but this should be a little bit more optimized because as you'll see below there is only a single call to `simps` instead of calling `simps` several times in a loop, for each redshift bin.
It's a small detail, but in practice it will lead to slightly more optimized code, because the computational graph will only contained a single integration op, that acts on a batch of dndz.